unit tests

msmkdenis opened the pull request.

msmkdenis committed:
texts: add unit tests

msmkdenis committed:
tests: naming

dragun-igor commented on precode_test.go:
```diff
@@ -0,0 +1,74 @@
+			assert.Equal(t, test.expectedCode, w.Code)
+			assert.Equal(t, test.expectedBody, w.Body.String())
+			if test.expectedCode == http.StatusOK {
+				assert.Equal(t, test.expectedCafeCount, len(strings.Split(w.Body.String(), ",")))
```
Отличный код! Но предлагаю попробовать ещё один кейс: сделай тест на корректный запрос с параметром count=0.

msmkdenis replied:
Согласен, такой тест напрашивался, но я его не добавлял, т.к. у нас в основном коде слишком слабые проверки.
Если count сделать -1 или 0 мы поймаем панику (out of bounds) или получим пустой ненужный ответ.
На мой взгляд тут корректно бы сделать проверку на positive значение count (пока так реализовал).
В качестве альтернативы можно сделать дефолтное значение 1.

dragun-igor replied:
Да, я согласен. Но раз пишем табличные тесты с единой логикой, хотелось бы чтобы они работали правильно во всех кейсах. Круто, что ты сам добавил проверку на положительность count в ручке.
Могу тогда разве что предложить изучить пакет [validator](https://pkg.go.dev/github.com/go-playground/validator/v10), который можно использовать для валидации входных данных при разработке АПИ. Хороший гибкий инструмент.

dragun-igor replied:
По теме тестирования также могу предложить изучить написание mock-тестов (пакет [gomock](https://pkg.go.dev/github.com/golang/mock/gomock), [mockery](https://github.com/vektra/mockery) или др.). Очень нужный навык на практике.
Также можно, на будущее, изучить [тестирование баз данных с помощью поднятия docker-контейнеров](https://habr.com/ru/articles/751220/). Это совсем advanced навык, тут важно умение работы с БД и docker. Но можно просто ознакомиться с такой практикой и попробовать её использовать, когда будут навыки. Если это слишком непонятно, то советую не зарываться и вернуться позже, ибо это совсем не по уровню спринта. Пакет [testcontainers-go](https://testcontainers.com/guides/getting-started-with-testcontainers-for-go/)

msmkdenis replied:
Спасибо! Очень полезные ссылки, обязательно разберусь!

dragun-igor reviewed (commented):
Привет, Денис. Меня зовут Игорь. Очень крутая работа. Хорошо описаны тест кейсы, код унифицированный. Но есть одно замечание, которое хотелось бы чтобы ты исправил.

msmkdenis committed:
fix: review 1 fix tests

add 2 testcases to test mainhandler with count 0 and -1
add check for count (positive) in mainhandler
fixed testcase in mainhandler: wrong city value

msmkdenis commented:
Привет, спасибо за ревью!
Поправил по комментарию, плюс нашел у себя "неправильный" тест (исправил))

dragun-igor commented on precode.go:
```diff
@@ -1,58 +1,45 @@
+	}
+
+	count, err := strconv.Atoi(countStr)
+	if err != nil || count <= 0 {
```
Эта проверка безусловно тут напршивается

dragun-igor reviewed (approved):
Привет! Работу принимаю!
